New shell API to get the hash

nethip opened the pull request:
This PR is heavily inspired by https://oroboro.com/unique-machine-fingerprint/ in trying to create a hash out of a system based on various parameters like network interfaces, volume, CPU e.t.c.

Relevant PR on Brackets side: https://github.com/adobe/brackets/pull/13419

nethip commented:
@abhijit @saurabh95 Please review this.

abhijitapte commented on appshell/appshell_extensions_gtk.cpp:
```diff
@@ -834,3 +848,190 @@ void DragWindow(CefRefPtr<CefBrowser> browser)
+{
+    u16 hash = 0;
+
+    for ( u32 i = 0; i < 6; i++ )
```
Where are these types u16 and u32 defined?

abhijitapte commented on appshell/appshell_extensions_gtk.cpp:
```diff
@@ -834,3 +848,190 @@ void DragWindow(CefRefPtr<CefBrowser> browser)
+    // get MAC address
+    bool foundMac1 = false;
+    struct ifreq* ifr;
+    for ( ifr = conf.ifc_req; (s8*)ifr < (s8*)conf.ifc_req + conf.ifc_len; ifr++ )
```
if the machine has multiple of them, what are we doing here?

nethip replied:
We are going to ignore this scenario @abhijitapte.

abhijitapte commented on appshell/appshell_extensions_gtk.cpp:
```diff
@@ -834,3 +848,190 @@ void DragWindow(CefRefPtr<CefBrowser> browser)
+    return id;
+}
+
+std::string GetSystemUniqueID()
```
can we use std::string& as a parameter instead?

nethip replied:
I don't think that is possible as I am returning a local variable here.

abhijitapte commented on appshell/appshell_extensions_mac.mm:
```diff
@@ -1675,3 +1695,160 @@ int32 GetArgvFromProcessID(int pid, NSString **argv)
+// we just need this for purposes of unique machine id. So any one or two
+// mac's is fine.
+
+u16 HashMacAddress( u8* mac )
```
I assume the mac code changes are the same as the ones mentioned for Linux.

nethip replied:
Yes more or less.

abhijitapte commented on appshell/appshell_extensions_platform.h:
```diff
@@ -59,6 +59,10 @@ static const int ERR_CL_TOOLS_NOTSUPPORTED  = 17;
 
 static const int ERR_PID_NOT_FOUND          = -9999; // negative int to avoid confusion with real PIDs
 
+typedef uint8_t   u8;
```
nevermind found the type definitions that I was looking for.

abhijitapte commented on appshell/appshell_extensions_win.cpp:
```diff
@@ -2005,6 +2009,129 @@ void DragWindow(CefRefPtr<CefBrowser> browser) {
+
+void GetMacHash(u16& mac1, u16& mac2)
+{
+	IP_ADAPTER_INFO AdapterInfo[32];
```
use camel case for adapterInfo

abhijitapte commented on appshell/appshell_extensions_win.cpp:
```diff
@@ -2005,6 +2009,129 @@ void DragWindow(CefRefPtr<CefBrowser> browser) {
+	IP_ADAPTER_INFO AdapterInfo[32];
+	DWORD dwBufLen = sizeof(AdapterInfo);
+
+	DWORD dwStatus = GetAdaptersInfo(AdapterInfo, &dwBufLen);
```
Should you consider MIB_IF_TYPE_ETHERNET as listed in the documentation https://msdn.microsoft.com/en-us/library/windows/desktop/aa365917(v=vs.85).aspx

nethip replied:
@abhijitapte Could you elaborate more on this comment?

abhijitapte commented on appshell/appshell_extensions_win.cpp:
```diff
@@ -2005,6 +2009,129 @@ void DragWindow(CefRefPtr<CefBrowser> browser) {
+
+	PIP_ADAPTER_INFO pAdapterInfo = AdapterInfo;
+	mac1 = HasMacAddress(pAdapterInfo);
+	if (pAdapterInfo->Next)
```
what should you assign mac2 to in case this condition doesn't hold?

nethip replied:
Actually it is better to initialize this all the ids to `0`.

abhijitapte commented on appshell/appshell_extensions_win.cpp:
```diff
@@ -2005,6 +2009,129 @@ void DragWindow(CefRefPtr<CefBrowser> browser) {
+	return hash;
+}
+
+u16 mask[5] = { 0x4e25, 0xf4a1, 0x5437, 0xab41, 0x0000 };
```
what is this mask for?

nethip replied:
I think this is used inside `Smear` method.

nethip commented:
@abhijitapte I will address the review comments. Thanks for reviewing this PR. Appreciate it!

nethip commented:
@abhijitapte I have addressed the review comments. Can you have a look again and tell me if this is OK now?

Note: I will be handling the linux change in a different PR.

nethip commented:
Raised PR for Linux side @ https://github.com/adobe/brackets-shell/pull/607

abhijitapte commented:
LGTM

nethip commented:
Thanks for reviewing this PR @abhijitapte

MarcelGerber commented:
For me on Windows, I cannot build brackets-shell due to `snprintf: identifier not found`.
Which VS version are you building with? I'm using VS 2012 Express.

nethip commented:
@MarcelGerber I am using VS 2015. It seems to be building fine on my system. I will try to build this again in the office on Monday. But thanks for building and letting me know. And good to hear from you after a long time 😁
